src/flag_gems/ops/avg_pool2d.py

```diff
@@ -0,0 +1,426 @@
+
+    h_in_offsets = h_block_idx * BLOCK_H + tl.arange(0, BLOCK_H)
+    w_in_offsets = w_block_idx * BLOCK_W + tl.arange(0, BLOCK_W)
+
+    grad_acc = tl.zeros((BLOCK_H, BLOCK_W), dtype=tl.float32)
+
+    for kh_loop in range(kernel_h):
+        for kw_loop in range(kernel_w):
+            h_out_num = h_in_offsets[:, None] + padding_h - kh_loop * dilation_h
+            w_out_num = w_in_offsets[None, :] + padding_w - kw_loop * dilation_w
+
+            h_valid_map = (h_out_num >= 0) & ((h_out_num % stride_h) == 0)
+            w_valid_map = (w_out_num >= 0) & ((w_out_num % stride_w) == 0)
+
+            h_out = h_out_num // stride_h
+            w_out = w_out_num // stride_w
+
+            h_out_mask = h_valid_map & (h_out < out_h)
+            w_out_mask = w_valid_map & (w_out < out_w)
+            out_mask = h_out_mask & w_out_mask
+
+            if divisor_override > 0:
+                divisor = tl.full(
+                    (BLOCK_H, BLOCK_W), divisor_override, dtype=tl.float32
+                )
+            elif COUNT_INCLUDE_PAD:
+                divisor = tl.full(
+                    (BLOCK_H, BLOCK_W), kernel_h * kernel_w, dtype=tl.float32
+                )
+            else:
+                # Re-compute count for the divisor when padding is not included.
+                h_start = h_out * stride_h - padding_h
+                w_start = w_out * stride_w - padding_w
+                count = tl.zeros((BLOCK_H, BLOCK_W), dtype=tl.int32)
+                for kh_count in range(0, kernel_h):
+                    for kw_count in range(0, kernel_w):
+                        h_in_for_count = h_start + kh_count * dilation_h
+                        w_in_for_count = w_start + kw_count * dilation_w
+                        is_valid = (
+                            (h_in_for_count >= 0)
+                            & (h_in_for_count < in_h)
+                            & (w_in_for_count >= 0)
+                            & (w_in_for_count < in_w)
+                        )
+                        count += is_valid.to(tl.int32)
+                divisor = count.to(tl.float32)
+
+            divisor = tl.where(divisor == 0, 1.0, divisor)
+
+            grad_out_ptr = (
+                grad_output_base_ptr + h_out * out_stride_h + w_out * out_stride_w
+            )
+            grad_out_val = tl.load(grad_out_ptr, mask=out_mask, other=0.0)
+            grad_to_add = grad_out_val / divisor
+            grad_acc += tl.where(out_mask, grad_to_add, 0.0)
+
+    grad_input_store_ptr = (
+        grad_input_block_ptr
+        + h_in_offsets[:, None] * in_stride_h
+        + w_in_offsets[None, :] * in_stride_w
+    )
+    in_write_mask = (h_in_offsets[:, None] < in_h) & (w_in_offsets[None, :] < in_w)
+    tl.store(
+        grad_input_store_ptr,
+        grad_acc.to(grad_input_ptr.type.element_ty),
+        mask=in_write_mask,
+    )
+
+
+class AvgPool2d(torch.autograd.Function):
```

Comment:
I suggest not to subclass `torch.autograd.Function` and add implementions for the forward and backward operator, separately.

- `avg_pool2d` or `avg_pool2d.out`
- `avg_pool2d_backward` and `avg_pool2d_backward.grad_input`

To do that, we need to check their schema in `pytorch/aten/src/ATen/native/native_functions.yaml` and their signature in the implementation files to be sure that our implementation uses the same signature.